x/operator/keeper/abci.go

```diff
@@ -2,116 +2,93 @@ package keeper
-	for assetID, priceChange := range priceChangeAssets {
-		// get the decimal of asset
-		assetInfo, err := k.assetsKeeper.GetStakingAssetInfo(ctx, assetID)
-		if err != nil {
-			return err
-		}
-		assetsDecimal[assetID] = assetInfo.AssetBasicInfo.Decimals
-		if _, ok := optedInAssetsInfo[assetID]; !ok {
-			optedInAssetsInfo[assetID] = make(map[string]string, 0)
-		}
-		// UpdateStateForAsset
-		f := func(assetID string, keys []string, state *operatortypes.OptedInAssetState) error {
-			newAssetUSDValue := CalculateShare(state.Amount, priceChange.NewPrice, assetInfo.AssetBasicInfo.Decimals, priceChange.Decimal)
-			changeValue := newAssetUSDValue.Sub(state.Value)
-			state.Value = newAssetUSDValue
 
-			avsAddr := keys[1]
-			avsOperator := string(types.GetJoinedStoreKey(keys[1], keys[2]))
-			AddShareInMap(shareChangeForAvsOperator, avsAddr, changeValue)
-			AddShareInMap(shareChangeForAvsOperator, avsOperator, changeValue)
-			optedInAssetsInfo[assetID][keys[2]] = avsAddr
-			return nil
-		}
-		err = k.IterateUpdateAssetState(ctx, assetID, f)
-		if err != nil {
-			return err
-		}
-	}
-	// BatchUpdateShareForAVSAndOperator
-	err = k.BatchUpdateShareForAVSAndOperator(ctx, shareChangeForAvsOperator)
+	// get the prices and decimals of assets
+	decimals, err := k.assetsKeeper.GetAssetsDecimal(ctx, assets)
 	if err != nil {
 		return err
 	}
-
-	// update the USD share for staker and operator
-	sharedParameter := &SharedParameter{
-		priceChangeAssets: priceChangeAssets,
-		assetsDecimal:     assetsDecimal,
-		optedInAssetsInfo: optedInAssetsInfo,
-		stakerShare:       make(map[string]sdkmath.LegacyDec, 0),
-	}
-	stakerShareHandleFunc := func(stakerID, assetID, operatorAddr string, state *delegationtype.DelegationAmounts) error {
-		UpdateShareOfStakerAndOperator(sharedParameter, assetID, stakerID, operatorAddr, state.UndelegatableAmount)
-		return nil
-	}
-	err = k.delegationKeeper.IterateDelegationState(ctx, stakerShareHandleFunc)
+	prices, err := k.oracleKeeper.GetMultipleAssetsPrices(ctx, assets)
 	if err != nil {
 		return err
 	}
-
-	operatorShareHandleFunc := func(operatorAddr, assetID string, state *types.OperatorAssetInfo) error {
-		UpdateShareOfStakerAndOperator(sharedParameter, assetID, "", operatorAddr, state.OperatorAmount)
+	// update the voting power of operators and AVS
+	avsVotingPower := sdkmath.LegacyNewDec(0)
+	opFunc := func(operator string, votingPower *sdkmath.LegacyDec) error {
+		// clear the old voting power for the operator
+		*votingPower = sdkmath.LegacyNewDec(0)
+		usdValue, selfUSDValue, err := k.CalculateUSDValueForOperator(ctx, operator, assets, decimals, prices)
+		if err != nil {
+			return err
+		}
+		// check if self USD value is more than the minimum self delegation.
+		minimumSelfDelegation, err := k.avsKeeper.GetAVSMinimumSelfDelegation(ctx, avsAddr)
+		if err != nil {
+			return err
+		}
+		if selfUSDValue.GTE(minimumSelfDelegation) {
+			*votingPower = votingPower.Add(usdValue)
```

Comment:
1. should `Add(selfUSDValue)` since `usdValue` is the total value of asset not specific operator
2. `votingPower` is not necessary, it equals to just add `avsVotingPower`

Reply:
No, the `selfUSDValue` is the value of assets owned by the operator itself, and the `usdValue` is the total value of a specific operator, including the assets delegated from the other stakers.
The `votingPower` is necessary, because the voting power of the operator will also be updated here. It's marked by a bool `isUpdate`